### ISSUE TITLE
Wayland Wrong Display Index

### ISSUE THREAD

**Issue by joshuarubin** [libsdl-org/SDL#4340]
I've got a wayland setup with 3 displays, one of which is a HIGHDPI (scaled) screen. I have `SDL_VIDEODRIVER=wayland`. I've found that `SDL_GetWindowPosition` always returns coordinates relative to display 0, not absolute coordinates that cover all displays. As a result `SDL_GetWindowDisplayIndex` always returns display 0.

This wouldn't be so bad except that my display 0 is not HIGHDPI, so when I start the app on my HIGHDPI screen (display 1), text is rendered extremely small.

`SDL_GetDisplayBounds` returns the correct values if I knew which display index to use for a window.

I'm using sway as my wayland window manager and these are not fullscreen windows.

This started out with something I found using kivy (https://github.com/kivy/kivy/issues/7499) but I've since reproduced it using just SDL2.

**Comment by flibitijibibo** [comment]
This should be fixed in the latest Git revision - you should see a minimum of one move event that puts the window at SDL_WINDOWPOS_CENTERED for the given display.

**Comment by joshuarubin** [comment]
This partially works. I do in fact get the display index updated with the git version of sdl2 (when waiting for, at lease one, move events). However, there is still a problem with `SDL_GL_GetDrawableSize` because when I move the window between monitors, it doesn't change. `SDL_GetRendererOutputSize` also has this issue.

**Comment by flibitijibibo** [comment]
This is the function you care about if you want to step through:

https://github.com/libsdl-org/SDL/blob/main/src/video/wayland/SDL_waylandopengles.c#L97

Here's where the scale is updated:

https://github.com/libsdl-org/SDL/blob/main/src/video/wayland/SDL_waylandwindow.c#L466

**Comment by joshuarubin** [comment]
It turns out that my simple test app wasn't rendering all that often and wasn't triggering opengl swapwindow, which in turn wasn't triggering the Wayland_HandlePendingResize. Once I fixed that, things are better, but still not perfect.

I'm able to detect switching between monitors and get the correct dpi for each monitor.

The two issues I have now are:
1. I have to manually adjust the render scale with a func like this (which I don't think should be necessary):
```c++
static void set_scale(SDL_Window *win) {
    SDL_Renderer *renderer = SDL_GetRenderer(win);
    int           px_w, px_h, w, h;
    SDL_GL_GetDrawableSize(win, &px_w, &px_h);
    SDL_GetWindowSize(win, &w, &h);
    float scale_w = static_cast<float>(px_w) / static_cast<float>(w);
    float scale_h = static_cast<float>(px_h) / static_cast<float>(h);
    SDL_RenderSetScale(renderer, scale_w, scale_h);
}
```
2. When moving windows between monitors, the contents often become corrupted until I resize the window. I've tried everything I can think of to fix this to no avail. Unsetting `$SDL_VIDEODRIVER` (which was `wayland`) gets things working (with xwayland) so I know my code _should_ be working. It's like I just can't trigger a full window refresh. The contents of my window shows a clock, so it's always changing and waiting doesn't help either.

**Comment by flibitijibibo** [comment]
For resize timing I think you may want #4326.

For render scaling you want SDL_RenderSetLogicalSize if you want scaling to be automatic, but aside from that this is very standard behavior and you should be rendering and presenting every frame.

**Comment by joshuarubin** [comment]
`SDL_RenderSetLogicalSize` was the missing piece. I tried the patch, but it turns out it wasn't needed. I now trigger this func whenever the window is moved/resized/size_changed:
```c++
static void set_scale(SDL_Window *win) {
    SDL_Renderer *renderer = SDL_GetRenderer(win);
    int           w, h;
    SDL_GetWindowSize(win, &w, &h);
    SDL_RenderSetLogicalSize(renderer, w, h);
}
```
This works great. Perfect scaling. Everything looks great when moving between screens. Thanks a bunch!

**Comment by flibitijibibo** [comment]
Glad it works! If there's nothing left to fix, this can be closed.

**Comment by joshuarubin** [comment]
Thanks for the help! Does that func look right to you, and when I trigger it?

**Comment by flibitijibibo** [comment]
That looks okay to me - if you want another example, VVVVVV also uses this call to scale its fixed-size viewport.

**Comment by joshuarubin** [comment]
Great. Thanks again!

TLDR I upgraded to sdl2 git/master and changed some of the way I used the library to fix this.
